src/main/scala/org/graphframes/lib/ConnectedComponents.scala

```diff
@@ -17,38 +17,262 @@
+    // TODO: This assignment job might fail if the graph is skewed.
+    val vertices = graph.indexedVertices
+      .select(col(LONG_ID).as(ID), col(ATTR))
+      // TODO: confirm the contract for a graph and decide whether we need distinct here
+      // .distinct()
+    val edges = graph.indexedEdges
+      .select(col(LONG_SRC).as(SRC), col(LONG_DST).as(DST))
+    val orderedEdges = edges.filter(col(SRC) !== col(DST))
+      .select(minValue(col(SRC), col(DST)).as(SRC), maxValue(col(SRC), col(DST)).as(DST))
+      .distinct()
+    GraphFrame(vertices, orderedEdges)
+  }
 
+  /**
+   * Returns the min neighbors of each vertex in a DataFrame with three columns:
+   *   - `src`, the ID of the vertex
+   *   - `min_nbr`, the min ID of its neighbors
+   *   - `cnt`, the total number of neighbors
+   *
+   * @return a DataFrame with three columns
+   */
+  private def minNbrs(ee: DataFrame): DataFrame = {
+    symmetrize(ee).groupBy(SRC).agg(min(col(DST)).as(MIN_NBR), count("*").as(CNT))
+  }
+
+  private def minValue(x: Column, y: Column): Column = {
+    when(x < y, x).otherwise(y)
+  }
+
+  private def maxValue(x: Column, y: Column): Column = {
+    when(x > y, x).otherwise(y)
+  }
+
+  /**
+   * Performs a possibly skewed join between edges and current component assignments.
+   * The skew join is done by broadcast join for frequent keys and normal join for the rest.
+   */
+  private def skewedJoin(
+      edges: DataFrame,
+      minNbrs: DataFrame,
+      broadcastThreshold: Int,
+      logPrefix: String): DataFrame = {
+    import edges.sqlContext.implicits._
+    val hubs = minNbrs.filter(col(CNT) > broadcastThreshold)
+      .select(SRC, MIN_NBR)
+      .as[(Long, Long)]
+      .collect()
+      .toMap // TODO: use OpenHashMap
+    if (hubs.isEmpty) {
+      return edges.join(minNbrs, SRC)
+    } else {
+      logger.debug(s"$logPrefix Number of skewed keys: ${hubs.size}.")
+    }
+    val isHub = udf { id: Long =>
+      hubs.contains(id)
+    }
+    val minNbr = udf { id: Long =>
+      hubs(id)
+    }
+    val hashJoined = edges.filter(!isHub(col(SRC)))
+      .join(minNbrs.filter(!isHub(col(SRC))), SRC)
+      .select(SRC, DST, MIN_NBR)
+    val broadcastJoined = edges.filter(isHub(col(SRC)))
+      .select(col(SRC), col(DST), minNbr(col(SRC)))
+    hashJoined.unionAll(broadcastJoined)
+  }
+
+  /**
+   * Runs connected components with default parameters.
+   */
   def run(graph: GraphFrame): DataFrame = {
-    val gx = graphxlib.ConnectedComponents.run(graph.cachedTopologyGraphX)
-    GraphXConversions.fromGraphX(graph, gx, vertexNames = Seq(COMPONENT_ID)).vertices
+    new ConnectedComponents(graph).run()
   }
 
-  private val COMPONENT_ID = "component"
+  private def run(graph: GraphFrame, broadcastThreshold: Int): DataFrame = {
+    val runId = UUID.randomUUID().toString.takeRight(8)
+    val logPrefix = s"[CC $runId]"
+    logger.info(s"$logPrefix Start connected components with run ID $runId.")
```

Comment:
`logger.info` uses macro so it doesn't unnecessarily materialize the message. I'm not sure whether `logInfo` behaves the same.